### PR TITLE
feat(dashboards): RFC for dashboards

### PR DESCRIPTION
**Pull request by keremsahin1**
## Checklist
- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [X] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Comment by jplaisted** [comment]
link to view in markdown https://github.com/linkedin/datahub/blob/dashboars_rfc/docs/rfcs/active/dashboards/1778-dashboards.md

File: docs/rfcs/active/dashboards/1778-dashboards.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+In this picture, `Dash_A` and `Dash_B` are dashboards and they are connected to charts through `Contains` edges.
+`C1`, `C2`, `C3` and `C4` are charts and they are connected to underlying datasets through `DownstreamOf` edges.
+`D1`, `D2` and `D3` are datasets.
+
+## How we teach this
+
+We should create/update user guides to educate users for:
+ - Search & discovery experience (how to find a dashboard in DataHub)
+ - Lineage experience (how to find upstream datasets of a dashboard and how to find dashboards generated from a dataset)
+
```
Comment (mars-lan) [inline docs/rfcs/active/dashboards/1778-dashboards.md:95]: Also include example crawlers

File: docs/rfcs/active/dashboards/1778-dashboards.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+## Alternatives
+
+In this design, there is an underlying assumption that input sources for charts will be a list of datasets.
```
Comment (mars-lan) [inline docs/rfcs/active/dashboards/1778-dashboards.md:98]: Can we at least make the upstream URN as a union so we can add metrics support later if we want?

Comment (jplaisted) [reply docs/rfcs/active/dashboards/1778-dashboards.md:98]: Would it make more sense to just have different lineage aspects in that case? On for datasets, and one for metrics, etc? Then you don't need a union?

Comment (mars-lan) [reply docs/rfcs/active/dashboards/1778-dashboards.md:98]: Possible, but you'll need to maintain multiple aspects & builders that look 90% identical. Not a fun place to be as the number of URN types grow.

File: docs/rfcs/active/dashboards/1778-dashboards.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+## Summary
+
+Adding support for dashboards (and charts) metadata cataloging and enabling search & discovery for them.
+The design should accomodate for different dashboarding ([Looker](www.looker.com), [Redash](www.redash.io)) tools used within a company.
+
+## Motivation
+
+Dashboards are a key piece within a data ecosystem of a company. They are used by different groups of employees across different organizations.
+They provide a way to visualize some data assets (tracking datasets or metrics) by allowing slice and dicing of the input data source.
+When a company scales, data assets including dashboards gets richer and bigger. Therefore, it's important to find and access to the right dashboard.
```
Comment (jplaisted) [inline docs/rfcs/active/dashboards/1778-dashboards.md:16]: Maybe you should explicitly have a list of requirements and non requirements? Access is pretty important, I think it is worth explicitly stating what we're trying to do here.

It seems like we're just trying to surface that information so it can be query-able (find all publish dashboards). We are NOT trying to enforce it. Additionally, we are NOT trying to build this so that a preview may eventually be embedded in the frontend, because there may be access restrictions.

tl;dr We don't want to be in the business of enforcing access to dashboards (yet, maybe ever?), which also means no previews.

File: docs/rfcs/active/dashboards/1778-dashboards.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
+- Start Date: 2020-08-03
```
Comment (jplaisted) [inline docs/rfcs/active/dashboards/1778-dashboards.md:1]: Quick question but do you think we should name the directory `1778-dashboards` too? Or do you think we should have these directory groupings? I'm fine either way, but it may make something harder to find if they're one level down.

(I'm not saying get rid of the directory, btw, I think this is good because you have images)

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 2 above)
Comment (jplaisted) [reply docs/rfcs/active/dashboards/1778-dashboards.md:98]: While not fun, it is more extensible, and helps prevents merge conflicts if these are your internal only aspects.

Comment (jplaisted) [reply docs/rfcs/active/dashboards/1778-dashboards.md:98]: Like, I wonder if our RFCs shouldn't have an `extensibility` section? :)

**Comment by liangjun-jiang** [comment]
not a direct review about this rfc. but a question where/how a rfc doc like this should be kept. It looks like @keremsahin1 created a new folder `rfcs/active` under `docs`. 
right now, we do have a folder `rfc` under `docs`. 
Since I am sending a new rfc. wonder how I should proceed, follow @keremsahin1 's way?

**Comment by jplaisted** [comment]
> not a direct review about this rfc. but a question where/how a rfc doc like this should be kept. It looks like @keremsahin1 created a new folder `rfcs/active` under `docs`.
> right now, we do have a folder `rfc` under `docs`.
> Since I am sending a new rfc pr, wonder how I should proceed, follow @keremsahin1 's way?

Ah, good catch. @keremsahin1 followed the documentation lol. I'll fix it.

@keremsahin1 can you move this to `docs/rfc/active` (remove the `s` in `rfcs`?)

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 4 above)
Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md:1]: Thanks for the suggestion, I think it would be better to have the directory name as `1778-dashboards`. And I guess, I can get rid of the `/imgs` subdirectory in that case. I will convert it to below structure:

```
1778-dashboards
|
 -- README.md
 -- dashboards_graph.png
 -- high_level_design.png
```

Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md:1]: Can you please update the RFC template/guide accordingly?

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 3 above)
Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md:16]: > Maybe you should explicitly have a list of requirements and non requirements?

I can convert to that structure but I believe RFC template should dictate the standard for the RFCs. Actually, template (and questions) helped me a lot while writing this.

> Is that assumption correct?

That's correct. I can add these.

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 1 above)
Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md:95]: Didn't quite get this. You mean crawlers as a separate section or into the "how we teach this" section?

Comment (mars-lan) [reply docs/rfcs/active/dashboards/1778-dashboards.md:95]: I mean education the users through example crawlers.

**Comment by keremsahin1** [comment]
> > not a direct review about this rfc. but a question where/how a rfc doc like this should be kept. It looks like @keremsahin1 created a new folder `rfcs/active` under `docs`.
> > right now, we do have a folder `rfc` under `docs`.
> > Since I am sending a new rfc pr, wonder how I should proceed, follow @keremsahin1 's way?
> 
> Ah, good catch. @keremsahin1 followed the documentation lol. I'll fix it.
> 
> @keremsahin1 can you move this to `docs/rfc/active` (remove the `s` in `rfcs`?)

Sure, I can change the directory name to `rfc`. Yes, I was following the documentation and thought `rfcs` is the correct directory name :)

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 1 above)
Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md:95]: My initial understanding was that this section is mainly to answer `How we teach this to end users of DataHub` not the `developers`? Example crawler guide sounds like a developer guide to me? @jplaisted Can you clarify what this section aims to cover?

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 3 above)
Comment (jplaisted) [reply docs/rfcs/active/dashboards/1778-dashboards.md]: > I can convert to that structure but I believe RFC template should dictate the standard for the RFCs. Actually, template (and questions) helped me a lot while writing this.

I agree. I can put together another PR for that.

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 4 above)
Comment (jplaisted) [reply docs/rfcs/active/dashboards/1778-dashboards.md]: Will do

File: docs/rfcs/active/dashboards/1778-dashboards.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+They might be using single or multiple data sources and generally have an associated query running against
+the underlying data source to generate the data that it will present.
+
+Below is a list of metadata which can be associated with a chart:
+
+- Title
+- Description
+- Type (Bar chart, Pie chart, Scatter plot etc.)
+- Input sources
+- Query (and its type)
```
Comment (pardhugunnam) [inline docs/rfcs/active/dashboards/1778-dashboards.md]: Similar to Input sources(with type  of input source embedded into definition), Query should also be a explicitly defined. We dont have Query as an entity as of now, but we can define some basic structure there.

File: docs/rfcs/active/dashboards/1778-dashboards.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ - Metric entity is not open source
+ - Popular dashboarding tools (Looker and Redash) generally use datasets as inputs directly and allow slicing & dicing 
+ on the dataset itself using their queries. In some way, they are generating the metrics on the fly.
+
+## Rollout / Adoption Strategy
+
+The design is supposed to be generic enough that any user of the DataHub should easily be able
+to onboard their dashboard metadata to DataHub irrespective of their dashboarding platform.
+
+Only thing users will need to do is to write an ETL script customized for their
```
Comment (pardhugunnam) [inline docs/rfcs/active/dashboards/1778-dashboards.md]: Do we intent to provide any sample crawlers?

File: docs/rfc/active/1778-dashboards/README.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+![high level design](high_level_design.png)
+
+As shown in the above diagram, dashboards are composed of a collection of charts at a very high level. These charts
+could be shared by different dashboards. In the example sketched above, `Chart_1`, `Chart_2` and `Chart_3` are part of
+`Dashboard_A` and `Chart_3` and `Chart_4` are part of `Dashboard_B`.
+
+### Entities
+There will be 2 top level GMA [entities](../../../what/entity.md) in the design: dashboards and charts.
+It's important to make charts as a top level entity because charts could be shared between different dashboards.
+We'll need to build `Contains` relationships between Dashboard and Chart entities.
```
Comment (jywadhwani) [inline docs/rfc/active/1778-dashboards/README.md:33]: One drawback of leveraging `Contains` relationship is that it is generic enough for any `Has-A` relationship and contains no additional metadata. But what if we want to attach additional metadata to this edge between Dashboard and Chart entities?

Comment (camelliazhang) [reply docs/rfc/active/1778-dashboards/README.md:33]: > One drawback of leveraging `Contains` relationship is that it is generic enough for any `Has-A` relationship and contains no additional metadata. But what if we want to attach additional metadata to this edge between Dashboard and Chart entities?

Extra attributes can be attached to a relationship, contains model can evolve in a backward compatible way.

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 6 above)
Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md]: We don't have any plans for this right now but maybe `Tableau` (if that's something we'll crawl at Linkedin going forward).

File: docs/rfc/active/1778-dashboards/README.md (hunk 7 above)
Comment (keremsahin1) [reply docs/rfc/active/1778-dashboards/README.md:33]: +1 to @camelliazhang. Edges can hold attributes as well.

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 5 above)
Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md]: We'll be defining a model for a `ChartQuery`. Please refer to https://github.com/linkedin/datahub/blob/a96bc672d3d6b27ee2679bfe9ea56e4d783f4753/metadata-models/src/main/pegasus/com/linkedin/chart/ChartQuery.pdl

How does its structure look like? Do you think of any additional metadata that we can capture for the query?

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 1 above)
Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md]: Ping @jplaisted for clarification on this.

File: docs/rfcs/active/dashboards/1778-dashboards.md (hunk 2 above)
Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md]: I guess we can change the type of inputs for a chart to a generic Urn [here](https://github.com/linkedin/datahub/pull/1775/files#diff-9e4e89860d537ee7aaf9cc4438291778R31). WDYT?

Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md]: But yeah, we can define a `ChartInput` enum with DatasetUrn (and MetricUrn later on) as well. That would be stricter and better, I guess.

Also, I'm not quite on the side of having duplicate aspects for this.

Comment (keremsahin1) [reply docs/rfcs/active/dashboards/1778-dashboards.md]: Refer to https://github.com/linkedin/datahub/pull/1775/files#diff-55f19bf92bba781d0244d6487dba29bd
